GBIF_Indo.R

```diff
@@ -25,113 +24,6 @@ bio_bone <- bio_bone[, !(names(bio_bone) %in% c('API'))]
-  return(grep("^[A-Za-z]+\\s(?!cf)[A-Za-z]+\\s[a-z]+$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-normal.species.rows <- function(value = FALSE) {
-  return(grep("^[A-Za-z]+\\s[A-Za-z]+(?:\\-[a-z]+)?$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-normal.genus.rows <- function(value = FALSE) {
-  return(grep("^[A-Za-z]+$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-species.rows.with.brackets <- function(value = FALSE) {
-  return(grep("^[A-Za-z]+\\s\\([A-Za-z]+\\)\\s[A-Za-z]+$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-species.rows.with.cfs <- function(value = FALSE) {
-  return(grep("^[A-Za-z]+\\scf\\.\\s[a-z]+$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-check.invalid.scientific.names <- function() {
-  # V and "ignore" abnormal cases will be placed here
-  na_rows <- which(is.na(bio_bone$scientificName))
-  print(paste('NA:', length(na_rows)))
-  print(paste('Subspecies:', length(normal.subspecies.rows())))
-  print(paste('Species:', length(normal.species.rows()) + length(species.rows.with.brackets()) + length(species.rows.with.cfs())))
-  print(paste('Genus:', length(normal.genus.rows())))
-  print('INVALID DATA:')
-  bio_bone$scientificName[-c(
-    na_rows,
-    normal.subspecies.rows(),
-    normal.species.rows(),
-    normal.genus.rows(),
-    species.rows.with.brackets(),
-    species.rows.with.cfs()
-  )]
-}
-check.invalid.scientific.names() # we know that 113 scientific names are still invalid
-
-
-# Cleaning 1
-# Check for anything ended with p. and . (dot) on scientificName field
-grep("^[A-Za-z ]+p\\.$", bio_bone$scientificName, perl=TRUE, value=TRUE)
-grep("^[A-Za-z ]+\\.$", bio_bone$scientificName, perl=TRUE, value=TRUE)
-# Remove p. and . (dot)
-bio_bone$scientificName <- trimws(gsub("^([A-Za-z ]+)p\\.$", '\\1', bio_bone$scientificName))
-bio_bone$scientificName <- trimws(gsub("^([A-Za-z ]+)\\.$", '\\1', bio_bone$scientificName))
-
-# Cleaning 2
-# Check "Ada 3 kata, kata ke-3 huruf besar" on scientificName field
-grep("^[A-Za-z]+\\s[a-z]+\\s[A-Z](?:[a-z]+)?$", bio_bone$scientificName, perl=TRUE, value=TRUE)
-# Remove all with capital letter on the third word
-bio_bone$scientificName <- trimws(gsub("^([A-Za-z]+\\s[a-z]+)\\s[A-Z](?:[a-z]+)?$", '\\1', bio_bone$scientificName))
-
-# Cleaning 3
-# TODO: here
-# Cleaning 5
-# TODO: here
-# Cleaning 6
-# TODO: here
-# Cleaning 8
-bio_bone$scientificName <- trimws(gsub("^([A-Za-z ]+)--$", '\\1', bio_bone$scientificName))
-# Cleaning 9
-# TODO: here
-# Cleaning 10
-bio_bone$scientificName <- trimws(gsub("^([A-Za-z]+\\s)cf(\\s[A-Za-z]+)$", '\\1cf.\\2', bio_bone$scientificName))
-cf.with.uppercase.species <- function(value = FALSE) {
-  return(grep("^[A-Za-z]+\\scf\\.?\\s[A-Z][a-z]+$", bio_bone$scientificName, perl=TRUE, value=value))
-}
-bio_bone$scientificName[cf.with.uppercase.species()] <- firstup(tolower((cf.with.uppercase.species(TRUE))))
-# Cleaning 11
-# TODO: here
-
 # TAXONRANK
```

Comment:
ini ga dipindahin juga? soalnya ini bagian dari data cleaning, ini yg ngecleaning taxonrank

disini species dan subspecies udah dibenerin taxonranknya, tinggal yg genus dan kingdom, yg kingdom gampang tinggal ambil normal kingdom rows, yg genus aga sulit karna kalo cuma deteksi 1 kata, maka bisajadi itu bukan genus, bisajadi famili or yg lain, kecuali kita punya list of genus

Reply:
yg todo fix cf taxonrank di bawah ini, coba kamu cek aja keluaran dari code di atas2nya, intinya sih itu yg cf masih beda2, ada yg species ada yg genus, harusnya langsung difill aja dengan "Species" karna kata kamu cf itu harusnya taxonranknya species, hehe